### PR TITLE
Add owner column to details view

### PR DESCRIPTION
**Pull request by tomasmatus**
Fixes: #580

- [x] How to display and sort when user/group is a number
- [x] test for user/group is a number

Sorting is currently as follows:
- directories first
- by user
- by group
- filename

I'd like to think about if we should sort current user first (as #580 mentions), I'm not sure if I like the idea but I want to check it out.

### CODE REVIEW

**Comment by tomasmatus** [comment]
current screenshot, font size is probably too big and should match font size of the "Modified" column rather than "Name" column
image

File: src/files-card-body.tsx [hunk 1]
```diff
@@ -47,6 +47,22 @@ const _ = cockpit.gettext;
 function compare(sortBy: Sort): (a: FolderFileInfo, b: FolderFileInfo) => number {
     const dir_sort = (a: FolderFileInfo, b: FolderFileInfo) => Number(b.to === "dir") - Number(a.to === "dir");
     const name_sort = (a: FolderFileInfo, b: FolderFileInfo) => a.name.localeCompare(b.name);
+    const owner_sort = (a: string | number | undefined, b: string | number | undefined) => {
+        if (a === undefined || b === undefined) {
+            return 0;
```
Comment (cockpituous) [inline src/files-card-body.tsx:52]: These 2 added lines are not executed by any test.

File: src/files-card-body.tsx [hunk 2]
```diff
@@ -47,6 +47,22 @@ const _ = cockpit.gettext;
...
+        }
+        if (typeof a === "string" && typeof b === "string") {
+            return a.localeCompare(b);
+        } else if (typeof a === "string" && typeof b === "number") {
```
Comment (cockpituous) [inline src/files-card-body.tsx:56]: This added line is not executed by any test.

File: src/files-card-body.tsx [hunk 3]
```diff
@@ -47,6 +47,22 @@ const _ = cockpit.gettext;
...
+            // TODO how to sort?
+            return -1;
+        } else if (typeof a === "number" && typeof b === "string") {
```
Comment (cockpituous) [inline src/files-card-body.tsx:59]: These 2 added lines are not executed by any test.

File: src/files-card-body.tsx [hunk 4]
```diff
@@ -47,6 +47,22 @@ const _ = cockpit.gettext;
...
+            // TODO how to sort?
+            return 1;
+        } else if (typeof a === "number" && typeof b === "number") {
+            return a - b;
```
Comment (cockpituous) [inline src/files-card-body.tsx:63]: These 3 added lines are not executed by any test.

File: src/files-card-body.tsx [hunk 5]
```diff
@@ -71,6 +87,12 @@ function compare(sortBy: Sort): (a: FolderFileInfo, b: FolderFileInfo) => number
         return (a, b) => dir_sort(a, b) || (perms(b) - perms(a)) || name_sort(a, b);
     case Sort.least_permissive:
         return (a, b) => dir_sort(a, b) || (perms(a) - perms(b)) || name_sort(a, b);
+    case Sort.owner_asc:
+        return (a, b) => dir_sort(a, b) || owner_sort(a.user, b.user) ||
+                owner_sort(a.group, b.group) || name_sort(a, b);
+    case Sort.owner_desc:
+        return (a, b) => dir_sort(a, b) || owner_sort(b.user, a.user) ||
+                owner_sort(b.group, a.group) || name_sort(b, a);
```
Comment (cockpituous) [inline src/files-card-body.tsx:95]: This added line is not executed by any test.

File: src/files-card-body.tsx (hunk 1 above)
Comment (cockpituous) [inline src/files-card-body.tsx:52]: These 2 added lines are not executed by any test.

File: src/files-card-body.tsx (hunk 2 above)
Comment (cockpituous) [inline src/files-card-body.tsx:56]: This added line is not executed by any test.

File: src/files-card-body.tsx (hunk 3 above)
Comment (cockpituous) [inline src/files-card-body.tsx:59]: These 2 added lines are not executed by any test.

File: src/files-card-body.tsx (hunk 4 above)
Comment (cockpituous) [inline src/files-card-body.tsx:63]: These 3 added lines are not executed by any test.

File: src/files-card-body.tsx (hunk 5 above)
Comment (cockpituous) [inline src/files-card-body.tsx:95]: This added line is not executed by any test.

**Comment by tomasmatus** [comment]
@garrett to sum up my design related questions:

When UID/GID is a number instead of account/group name I think I should just show the number instead of doing something special. And when it comes to sorting I think we can get away with giving numbers higher priority and sort them as numbers, then local users/groups which are sorted with localeCompare.

Second question is about fontsize in the `Owner` column. Should it match font in `Modified` column or `Filename` column as it does now?

And lastly about "sorting current user first" I personally don't like this as to me it seems more confusing than useful but I am fine with waiting for your take on it.

**Comment by allisonkarlitskaya** [comment]
Writing down the result of an IRC conversation here for posterity:

We considered having separate user and group columns, and also considered always showing `user:group` within the same column, complete with different mixes of things like:

```
   oneuser : onegroup   
 otheruser : othergroup 
```

or

```
oneuser     onegroup
otheruser othergroup
```

but the main thrust of this is that in 99% of the directories that most users are likely to be browsing around, file ownership is the same for every file in the directory, and usually the group and username are the same.  In that case, it makes things a lot visually cleaner to leave out the group.

And yes, this looks weird and becomes difficult to read in non-uniform directories, but that case is rare.

Other ideas that are more difficult to implement:
 - only display the `:group` name if the group is different from the username, but then do it for every item in the directory.  ie: as soon as a single item with a different group appears, then all of them get the groupname added.  This is a bit "spooky action at a distance" and doesn't fit well with how react likes to do things...
 - if the uid/gid of all files in a directory are equal to the uid/gid of the directory itself then omit the column entirely.  Same arguments apply as with above (although I consider it slightly less evil).

File: src/files-card-body.tsx (hunk 1 above)
Comment (cockpituous) [inline src/files-card-body.tsx:52]: These 2 added lines are not executed by any test.

File: src/files-card-body.tsx (hunk 2 above)
Comment (cockpituous) [inline src/files-card-body.tsx:56]: This added line is not executed by any test.

File: src/files-card-body.tsx (hunk 3 above)
Comment (cockpituous) [inline src/files-card-body.tsx:59]: These 2 added lines are not executed by any test.

File: src/files-card-body.tsx (hunk 4 above)
Comment (cockpituous) [inline src/files-card-body.tsx:63]: These 3 added lines are not executed by any test.

File: src/files-card-body.tsx (hunk 5 above)
Comment (cockpituous) [inline src/files-card-body.tsx:95]: This added line is not executed by any test.

File: src/files-card-body.tsx (hunk 1 above)
Comment (cockpituous) [inline src/files-card-body.tsx:52]: These 2 added lines are not executed by any test.

Comment (cockpituous) [inline src/files-card-body.tsx:52]: These 2 added lines are not executed by any test.

**Comment by garrett** [comment]
> When UID/GID is a number instead of account/group name I think I should just show the number instead of doing something special. And when it comes to sorting I think we can get away with giving numbers higher priority and sort them as numbers, then local users/groups which are sorted with localeCompare.

What do you mean by "something special"? Do you have an example?

> Second question is about fontsize in the Owner column. Should it match font in Modified column or Filename column as it does now?

All the details should use the same font family and sizing, unless they're monospace, like the permissions. And numbers should be tabular.

(Modified date/time isn't to spec, which is probably throwing off your perception a bit. It's supposed to show the date, unless it's today, then it's the time, or yesterday, as then it says "Yesterday" and the time. Having a much longer string there means there's more to compare everything else to visually.)

> And lastly about "sorting current user first" I personally don't like this as to me it seems more confusing than useful but I am fine with waiting for your take on it.

I said "Current user first? (Not sure)" as an idea, not that we should do it, hence the "?" and "(Not sure)"

It should definitely sort by:

1. user account first
2. group second (that is: don't ignore the group when sorting, as it could wind up "random")
3. file/dir name third (as many files and dirs will have the same username and group)

…but that can be pure alphanumeric sorting.

**Comment by tomasmatus** [comment]
> What do you mean by "something special"? Do you have an example?

I don't have any specific ideas. I only saw dolphin shows empty entry and that doesn't fit our design.

> I said "Current user first? (Not sure)" as an idea, not that we should do it, hence the "?" and "(Not sure)"

Yes, I know that. I only wanted to share why I chose not to do it.

File: src/files-card-body.tsx (hunk 1 above)
Comment (cockpituous) [inline src/files-card-body.tsx:52]: These 2 added lines are not executed by any test.

File: src/header.tsx [hunk 6]
```diff
@@ -88,6 +90,17 @@ export const filterColumns = [
             label: _("Last modified"),
         },
     },
+    {
+        title: _("Owner"),
+        [SortByDirection.asc]: {
+            itemId: Sort.owner_asc,
+            label: _("Owner ascending"),
```
Comment (tomasmatus) [inline src/header.tsx:97]: @jelly oh yeah Garrett never commented on this and the `Owner descending` string. I am open for a discussion if you'd like to change it :)

**Review by jelly — COMMENTED** [review]
Overall LGTM, needs pixel tests and maybe @allisonkarlitskaya wants to do a check as they commented on the PR?

File: src/files-card-body.tsx [hunk 7]
```diff
@@ -505,6 +532,20 @@ const FilePermissions = ({ file } : {
...
+    const ownerText = (file.user === file.group)
+        ? file.user
+        : `${file.user}:${file.group}`;
+
+    return (
+        <TextContent>
+            <Text>
+                {ownerText}
+            </Text>
+        </TextContent>
```
[the remark this answers is not in the document]
Comment (tomasmatus) [reply src/files-card-body.tsx:545]: probably but it would need some extra css as it has a different fontsize

File: src/files-card-body.tsx (hunk 1 above)
Comment (cockpituous) [inline src/files-card-body.tsx:52]: These 2 added lines are not executed by any test.

**Review by allisonkarlitskaya — APPROVED** [review]

**Comment by allisonkarlitskaya** [comment]
> Overall LGTM, needs pixel tests and maybe @allisonkarlitskaya wants to do a check as they commented on the PR?

I don't think this really needs pixel tests?  This stuff is highly visible in many existing screenshots (the PR updates 20 of them)... Or do you mean for different sorting possibilities?

**Comment by jelly** [comment]
> I think what Jelle meant is that it needed pixeltests update

:100:  this :)

**Review by jelly — APPROVED** [review]

**Pull request merged**